### PR TITLE
🌱 KCP: cache secrets between LookupOrGenerate and ensureCertificatesOwnerRef

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #8814

### CODE REVIEW

**Comment by sbueringer** [comment]
/assign @fabriziopandini @killianmuldoon @ykakarap 

/hold
for
/test pull-cluster-api-e2e-full-main

File: controlplane/kubeadm/internal/controllers/controller.go [hunk 1]
```diff
@@ -928,38 +928,32 @@ func (r *KubeadmControlPlaneReconciler) adoptOwnedSecrets(ctx context.Context, k
...
+func (r *KubeadmControlPlaneReconciler) ensureCertificatesOwnerRef(ctx context.Context, certificates secret.Certificates, owner metav1.OwnerReference) error {
 	for _, c := range certificates {
-		s := &corev1.Secret{}
-		secretKey := client.ObjectKey{Namespace: clusterKey.Namespace, Name: secret.Name(clusterKey.Name, c.Purpose)}
-		if err := r.Client.Get(ctx, secretKey, s); err != nil {
-			return errors.Wrapf(err, "failed to get Secret %s", secretKey)
-		}
-
-		patchHelper, err := patch.NewHelper(s, r.Client)
+		patchHelper, err := patch.NewHelper(c.Secret, r.Client)
```
Comment (fabriziopandini) [inline controlplane/kubeadm/internal/controllers/controller.go:933]: nit: can we check if the secret already has the ownerRef and continue (without even creating the patch helper)

**Review by fabriziopandini — COMMENTED** [review]
/lgtm

File: controlplane/kubeadm/internal/controllers/controller.go (hunk 1 above)
Comment (killianmuldoon) [inline controlplane/kubeadm/internal/controllers/controller.go:933]: Should `c.Secret` be nil-checked before this - even though I don't think this situation can actually happen.

**Review by killianmuldoon — COMMENTED** [review]
One nit. If I understand right the `Get` on Secret is always uncached because Secrets aren't cached in KCP - so this change instead stores the secret somewhere we have ready access to it.

File: controlplane/kubeadm/internal/controllers/controller.go (hunk 1 above)
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller.go:937]: Thx, done!

**Comment by sbueringer** [comment]
> One nit. If I understand right the `Get` on Secret is always uncached because Secrets aren't cached in KCP - so this change instead stores the secret somewhere we have ready access to it.

Yes exactly. That's the idea. Just a small optimization because actually caching the secrets is not that trivial (because currenlty there is no way to identify which secrets exactly to cache and we don't want to cache all).

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

feel free to un-hold anytime

**Comment by sbueringer** [comment]
Thx for the reviews!

/hold cancel

**Pull request merged**